core/src/main/scala/kafka/server/KafkaRequestHandler.scala

```diff
@@ -108,6 +110,19 @@ class BrokerTopicMetrics(name: Option[String]) extends KafkaMetricsGroup {
   val failedFetchRequestRate = newMeter("FailedFetchRequestsPerSec", "requests", TimeUnit.SECONDS, tags)
   val totalProduceRequestRate = newMeter("TotalProduceRequestsPerSec", "requests", TimeUnit.SECONDS, tags)
   val totalFetchRequestRate = newMeter("TotalFetchRequestsPerSec", "requests", TimeUnit.SECONDS, tags)
+
+  val partitions = new ConcurrentHashMap[Int, Partition]
```

Comment:
@hachikuji `ReplicaManager` which calls `onPartitionAdd()` and `onPartitionDelete()` is thread-safe and handles concurrent calls to add/remove using a ConcurrentHashMap. A counter or a list of partitions that relies just on topic/partition-id doesn't provide the same thread safety. It has been a long time since I wrote the code, I will take a look again to see if the code can be simplified.